crates/vm/levm/src/precompiles.rs

```diff
@@ -501,12 +501,186 @@ pub fn ecmul(
     }
 }
 
-fn ecpairing(
-    _calldata: &Bytes,
-    _gas_for_call: u64,
-    _consumed_gas: &mut u64,
+pub fn ecpairing(
+    calldata: &Bytes,
+    gas_for_call: u64,
+    consumed_gas: &mut u64,
 ) -> Result<Bytes, VMError> {
-    Ok(Bytes::new())
+    // The input must always be a multiple of 192 (6 32-byte values)
+    if calldata.len() % 192 != 0 {
+        return Err(VMError::PrecompileError(PrecompileError::ParsingInputError));
+    }
+
+    let groups_number = calldata.len() / 192;
+
+    // Consume gas
+    let gas_cost = gas_cost::ecpairing(groups_number)?;
+    increase_precompile_consumed_gas(gas_for_call, gas_cost, consumed_gas)?;
+
+    let mut mul: FieldElement<Degree12ExtensionField> = QuadraticExtensionFieldElement::one();
+    for group_number in 0..groups_number {
+        // Define the group indexes and slice calldata to get the group data
+        let group_start = group_number
+            .checked_mul(192)
+            .ok_or(InternalError::ArithmeticOperationOverflow)?;
+        let next_group_start = (group_number
```

Comment:
Isn't this just `group_start + 192`?

Reply:
In this context you may want to call it `group_end` or `input_end`, I know it is the beginning of the next group but you use it for getting the calldata between the start and end of the group

Reply:
Done in [Change group index calculation logic and name](https://github.com/lambdaclass/ethrex/pull/1560/commits/cecff1d307e923a584621383577ac07951353752)